examples/platforms/efr32mg12/flash.c

```diff
@@ -28,22 +28,373 @@
 
 /**
  * @file
- *   This file implements the OpenThread platform abstraction for the non-volatile storage.
+ *   This file implements the OpenThread platform abstraction for the non-volatile
+ *   storage for the EFR32 platform using either the Silabs Nvm3 interface or
+ *   the Nest developed nv interface.
  */
 
-#include <openthread-core-config.h>
-
 #include <openthread/config.h>
-#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/settings.h>
 
-#include "utils/code_utils.h"
-#include "utils/flash.h"
+#if OPENTHREAD_USE_THIRD_PARTY_NVM_MANAGER
+
+#include "nvm3.h"
+#include "nvm3_hal_flash.h"
+#include <string.h>
+#include "common/code_utils.hpp"
+
+// Two macros are provided to support the creation of the Silicon Labs NVM3 area and
+// initialization data- NVM3_DEFINE_SECTION_STATIC_DATA() and NVM3_DEFINE_SECTION_INIT_DATA().
+// A linker section called 'name'_section is defined by NVM3_DEFINE_SECTION_STATIC_DATA().
+// The NVM3 area is placed at the top of the device FLASH section by the linker script
+// file: e.g. efr32mgxx.ld
+// An error is returned by nvm3_open() on alignment or size violation.
+
+// Local version of SDK macro (avoids uninitialized var compile error).
+#define OT_NVM3_DEFINE_SECTION_STATIC_DATA(name, nvmSize, cacheSize) \
+    static nvm3_CacheEntry_t name##_cache[cacheSize];                \
+    static uint8_t           name##_nvm[nvmSize] SL_ATTRIBUTE_SECTION(STRINGIZE(name##_section))
+
+// Local version of SDK macro (allows OT to configure the maximum nvm3 object size and headroom).
+#define OT_NVM3_DEFINE_SECTION_INIT_DATA(name, maxObjectSize, repackHeadroom) \
+    static nvm3_Init_t name = {                                               \
+        (nvm3_HalPtr_t)name##_nvm,                                            \
+        sizeof(name##_nvm),                                                   \
+        name##_cache,                                                         \
+        sizeof(name##_cache) / sizeof(nvm3_CacheEntry_t),                     \
+        maxObjectSize,                                                        \
+        repackHeadroom,                                                       \
+        &nvm3_halFlashHandle,                                                 \
+    }
+
+#define NUM_SETTINGS_OBJECTS 7 // == number of enumerated Settings key types (in settings.hpp).
+#define NUM_INDEXED_SETTINGS \
+    OPENTHREAD_CONFIG_MLE_MAX_CHILDREN // Indexed key types are only supported for kKeyChildInfo (=='child table').
```

Comment:
Will things still work in the scenario where a firmware update changes the value of `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN` (either smaller or larger)?

Reply:
This will still work because OPENTHREAD_CONFIG_MLE_MAX_CHILDREN  is used to calculate the size of the ram cache that the nvm3 system uses for fast access - there is no change to object addressing.